### PR TITLE
design

### PR DESCRIPTION
**Pull request by gravesti**
# Pull Request

I've put together a design for how we can refactor the model creation code to make the package more flexibl.
I have a proposal to make `make_model_string_model()` into a generic that takes the outcome and borrowing type, so we can flexibly create add new methods without having to worry about all kinds of special cases.

### CODE REVIEW

**Review by danielinteractive — COMMENTED** [review]
thanks @gravesti ! I really like the class/methods based design proposal. That also makes it possible (or we should try to make it work) to extend the package functionality without changing package code.

File: vignettes/dataset.Rmd [hunk 1]
```diff
@@ -321,18 +321,17 @@ variable we are interested in by plotting histograms of the draws with the
 function `mcmc_hist()`. Let's do that for the Hazard ratio for the treatment
 effect and for our commensurability parameter, tau.
 
-```{r mcmc-hist, message = FALSE}
-library(bayesplot)
-mcmc_hist(draws, c("treatment HR", "commensurability parameter"))
+```{r mcmc-hist, message = FALSE, eval = requireNamespace("bayesplot")}
```
Comment (danielinteractive) [inline vignettes/dataset.Rmd:324]: ah cool stuff, did not see this trick before

**Comment by mattsecrest** [comment]
@gravesti yeah I like this approach and agree it is more extensible to new methods. Let's discuss briefly soon and then push updates in our current methods before adding in new methods.

**Comment by danielinteractive** [comment]
@mattsecrest Cool - and since this is a design doc feel free to merge earlier so that you have this PR out of the way (one design is aligned)

**Comment by mattsecrest** [comment]
@danielinteractive out of curiosity is this a common convention to keep `.Rmd` in `design`? I have not seen yet, but am happy to merge this to remove the PR as you suggest.

**Comment by danielinteractive** [comment]
Yes, e.g. see https://github.com/openpharma/mmrm/tree/main/design but also many other packages we use this. It can be quite useful even later on when revisiting design decisions made in the past.

**Review by mattsecrest — APPROVED** [review]

**Pull request merged**